tests/tests_fabric/strategies/test_model_parallel_integration.py

```diff
@@ -0,0 +1,489 @@
+
+
+class FeedForward(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.w1 = nn.Linear(32, 64)
+        self.w2 = nn.Linear(32, 64)
+        self.w3 = nn.Linear(64, 32)
+
+    def forward(self, x):
+        return self.w3(F.silu(self.w1(x)) * self.w2(x))
+
+
+def _parallelize_feed_forward_tp(model, device_mesh):
+    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
+
+    tp_mesh = device_mesh["tensor_parallel"]
+    tp_plan = {
+        "w1": ColwiseParallel(),
+        "w2": ColwiseParallel(),
+        "w3": RowwiseParallel(),
+    }
+    parallelize_module(model, tp_mesh, tp_plan)
+    return model
+
+
+def _parallelize_feed_forward_fsdp2(model, device_mesh):
+    from torch.distributed._composable.fsdp.fully_shard import fully_shard
+    from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import checkpoint_wrapper
+
+    dp_mesh = device_mesh["data_parallel"]
+    assert dp_mesh.ndim == 1  # Hybrid-sharding not supported
+
+    # Fully-shard each layer
+    fully_shard(model.w1, mesh=dp_mesh)
+    fully_shard(model.w2, mesh=dp_mesh)
+    fully_shard(model.w3, mesh=dp_mesh)
+
+    # Activation checkpointing
+    model = checkpoint_wrapper(model)
+
+    return model
+
+
+def _parallelize_feed_forward_fsdp2_tp(model, device_mesh):
+    model = _parallelize_feed_forward_tp(model, device_mesh)
+    model = _parallelize_feed_forward_fsdp2(model, device_mesh)
+    return model
+
+
+@RunIf(min_torch="2.3", standalone=True, min_cuda_gpus=2)
+def test_tensor_parallel():
+    from torch.distributed._tensor import DTensor
+
+    strategy = ModelParallelStrategy(parallelize_fn=_parallelize_feed_forward_tp)
+    fabric = Fabric(accelerator="auto", devices=2, strategy=strategy)
+    fabric.launch()
+
+    fabric.seed_everything(0)
+
+    with fabric.init_module(empty_init=True):
+        model = FeedForward()
+
+    optimizer = torch.optim.AdamW(model.parameters())
+    model, optimizer = fabric.setup(model, optimizer)
+
+    assert all(isinstance(weight, DTensor) for weight in model.parameters())
+    assert model.w1.weight.device_mesh == fabric.strategy.device_mesh["tensor_parallel"]
+
+    dataset_size = 6
+    dataset = RandomDataset(32, dataset_size)
+    dataloader = DataLoader(dataset, batch_size=2)
+    dataloader = fabric.setup_dataloaders(dataloader)
+
+    # No data sharding, all GPUs get the same input inside a TP group
+    assert len(dataloader) == dataset_size // dataloader.batch_size
+    assert isinstance(dataloader.sampler, DistributedSampler)
+
+    for _, batch in enumerate(dataloader):
+        # All batches must be identical across TP group
+        batches = fabric.all_gather(batch)
+        assert all(torch.equal(batches[0], batches[i]) for i in range(1, len(batches)))
+
+        output = model(batch)
+        fabric.backward(output.sum())
+        optimizer.step()
+        optimizer.zero_grad()
```

Comment:
Since this is already spending time on running fwd/bwd/opt.step you might as well add some sanity checks that things got updated as expected, or that shard shapes are as expected.

Same suggestion for the test below